### PR TITLE
Support for defaultSource

### PR DESCRIPTION
**Pull request by danielgindi**
Both iOS and Android.
No support for URIs due to library limitations. It also does not make any sense. (As defaultSource should be available immediately, not for going to fetch from a remote location)

### CODE REVIEW

**Comment by danielgindi** [comment]
@DylanVann Can you take a look at this?

**Comment by qalqi** [comment]
Does this act as fallback image as well In case of error loading image ?

**Comment by danielgindi** [comment]
> Does this act as fallback image as well In case of error loading image ?

Yes

**Comment by danielgindi** [comment]
Codecov fails on code specific to Android. Seems like the tests are not running on Android.

**Comment by danielgindi** [comment]
@DylanVann This solves the one big issue with this library, would be helpful to review. Thanks 👍

**Comment by geroale** [comment]
Hello! Thank you so much for this feature: it is very important also for me and my team, as well as for the overall project. Is it possible to fix or merge it in the master?
@DylanVann

**Comment by itsam** [comment]
Any news on this? Thanks.

**Comment by gvsakhil** [comment]
having defaultSource makes fastImage the best plugin :(

**Comment by ramonvermeulen** [comment]
I kind of really need this, any chance this can get merged/released in the near future?
Would be really thankful!

**Comment by danielgindi** [comment]
@DylanVann Any chance to get some attention here?

**Comment by danielgindi** [comment]
I've added more tests to satisfy codecov. Hope it's silenced now.

**Comment by enahum** [comment]
@danielgindi don't you think `defaultSource` should also support network images? I can easily see this as a way to load a thumbnail (small kb) while the full image is being downloaded

**Comment by danielgindi** [comment]
@enahum Well theoretically you would be correct, but practically no.  
If you have such a need - you are most probably doing something wrong:

* In most cases, such a feature is used in order to have something visible *immediately*, which means you have to be able to load the image synchronously, and have the dimensions available upfront, and ideally the image should be in the exact dimensions for display to avoid a costly resize operation.

* If you have the same default image for all images (most common case), then the image should just be local in your app.

* If you generate the default on the fly (like letters avatar) - it's still local, in memory or cached on disk.

* Some native libraries have an option to supply separate source for error image. In that case - you probably have a common error image to show for every error. This should be local in your app.

* If you are showing an image in a list view - then you want it to load fast and be available immediately, for smooth scrolling and to avoid visual glitches.

Now if we care about conforming to the built-in RN `Image` - then the `Image` docs say the image should be local or a static image resource. https://reactnative.dev/docs/image.

**Comment by marf** [comment]
Can this be merged with the master, please?
@danielgindi @DylanVann

**Comment by geroale** [comment]
Is it possible to finally merge this into master? Thank you in advance @danielgindi @DylanVann

**Comment by SauloDesenvolvedor** [comment]
defaultSource up!

**Comment by julioxavierr** [comment]
+1

**Comment by danielgindi** [comment]
@DylanVann I've rebased on `master` to resolve conflicts. The CI has an issue that's not related to the code (fails on `master` too).

File: src/index.tsx [hunk 1]
```diff
@@ -79,7 +83,8 @@ export interface ImageStyle extends FlexStyle, TransformsStyle, ShadowStyleIOS {
 }
 
 export interface FastImageProps {
-    source: Source | number
+    source: Source | ImageRequireSource
+    defaultSource: ImageRequireSource
```
Comment (vinayr) [inline src/index.tsx:87]:
```suggestion
    defaultSource?: ImageRequireSource
```

File: src/index.tsx [hunk 2]
```diff
@@ -143,6 +149,28 @@ function FastImageBase({
     ...props
 }: FastImageProps & { forwardedRef: React.Ref<any> }) {
     const resolvedSource = Image.resolveAssetSource(source as any)
+    let resolvedDefaultSource:
+        | ImageRequireSource
+        | string
+        | null = defaultSource
+
+    if (isAndroid) {
```
Comment (vinayr) [inline src/index.tsx:157]:
```suggestion
        | typeof defaultSource
        | string
        | null = defaultSource

    if (isAndroid && defaultSource) {
```
This fixes typescript build errors

Comment (danielgindi) [reply src/index.tsx:157]: There was a `&& defaultSource` there but TS complained. Did you test these changes?

Comment (danielgindi) [reply src/index.tsx:157]: The check should stay `if (isAndroid) else ...`. First case is for Android, `else` is for iOS. In iOS they are represented as numbers.

Comment (vinayr) [reply src/index.tsx:157]: The problem is `defaultSource` is optional, its value can be `undefined`. But `Image.resolveAssetSource` input parameter cannot be `undefined`.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-native/index.d.ts
```
resolveAssetSource(source: ImageSourcePropType)
export type ImageSourcePropType = ImageURISource | ImageURISource[] | ImageRequireSource;
export type ImageRequireSource = number;
```

Therefore I added `&& defaultSource` to if condition. As an alternative we can also do like this - 
```
if (!defaultSource) {
    resolvedDefaultSource = null
} else if (isAndroid) {
    ....
} else if (typeof resolvedDefaultSource !== 'number') {
    ....
}
```

**Comment by tarouboy** [comment]
@danielgindi 
Does it support v6.1.1?
I can't upgrade RN to 0.60+ in a short time but would like to use defaultSource :(

File: android/src/main/java/com/dylanvann/fastimage/FastImageViewWithUrl.java [hunk 3]
```diff
@@ -1,14 +1,129 @@
...
+                    // This will make this work for remote and local images. e.g.
+                    //    - file:///
+                    //    - content://
+                    //    - res:/
+                    //    - android.resource://
+                    //    - data:image/png;base64
+                    .load(imageSource == null ? null : imageSource.getSourceForLoad())
+                    .placeholder(mDefaultSource) // show until loaded
+                    .fallback(mDefaultSource) // null will not be treated as error
+                    .apply(FastImageViewConverter.getOptions(context, imageSource, mSource));
```
Comment (vinayr) [inline android/src/main/java/com/dylanvann/fastimage/FastImageViewWithUrl.java:121]: @danielgindi the local defaultSource image was not showing (tested on multiple emulators/devices and RN 0.61.5/0.62.2). Had to make following changes to make it work. Could you take a look?

```suggestion
                    .apply(FastImageViewConverter
                            .getOptions(context, imageSource, mSource)
                            .placeholder(mDefaultSource) // show until loaded
                            .fallback(mDefaultSource)); // null will not be treated as error              
```

Comment (danielgindi) [reply android/src/main/java/com/dylanvann/fastimage/FastImageViewWithUrl.java:121]: Could it be due to an update to the underlying library?
I have not seen such an issue, will have to test...

Comment (mralj) [reply android/src/main/java/com/dylanvann/fastimage/FastImageViewWithUrl.java:121]: I have had same issue on Android. 
Suggestion made by @vinayr fixed the problem for me as well :) 
According to SO [link](https://stackoverflow.com/questions/45481756/glide-4-0-0-missing-placeholder-error-glideapp-and-does-not-resolve-its-method) & [link](https://stackoverflow.com/questions/47262017/glide-4-3-1-override-and-placeholder-features-not-work) this is @vinayr's suggestion is way to go!

**Comment by neuneed** [comment]
Any chance to merge this? @DylanVann @danielgindi 
Seems there was no any other issues in this branch?

**Comment by danielgindi** [comment]
@neuneed Yes it had some work done on it, seems stable even after latest changes in master. But there's no response from @DylanVann. I also emailed him, but again not response.

**Comment by marf** [comment]
Hope this will be merged soon on the master @DylanVann 
Thank you!

File: android/src/main/java/com/dylanvann/fastimage/FastImageViewWithUrl.java (hunk 3 above)
Comment (neuneed) [reply android/src/main/java/com/dylanvann/fastimage/FastImageViewWithUrl.java]: @danielgindi I have just tested your branch in real device for both iOS and Android, so turns out for defaultSource in iOS is totally working, but in Android device seems not. So if the Glide api changed maybe should update it.

**Comment by archansel** [comment]
Waiting for this merge too

**Comment by deepgosalia1** [comment]
The default source prop is up ?? its really needed in my project as well! would mean a ton!

**Comment by vineethvijayan** [comment]
Waiting for this

**Comment by varunkv2019** [comment]
eagerly waiting for the feature

**Comment by rpavlovs** [comment]
@danielgindi could you rebase again, please 🙏

**Comment by rpavlovs** [comment]
Installing from this PR directly I get the following error from bundler:
```
error: Error: While trying to resolve module `react-native-fast-image` from file `.../TabNavigator.tsx`, the package `.../node_modules/react-native-fast-image/package.json` was successfully found. 
However, this package itself specifies a `main` module field that could not be resolved (`.../node_modules/react-native-fast-image/dist/index.js`. 
Indeed, none of these files exist:
  * .../node_modules/react-native-fast-image/dist/index.js(.native|.ios.js|.native.js|.js|.ios.json|.native.json|.json|.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.svg|.native.svg|.svg)
  * .../node_modules/react-native-fast-image/dist/index.js/index(.native|.ios.js|.native.js|.js|.ios.json|.native.json|.json|.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.svg|.native.svg|.svg)
```

**Comment by danielgindi** [comment]
@rpavlovs I've rebased. Please anyone test this as I haven't tested it with the latest `master` changes.

Also, @DylanVann I'm getting tired of this. Could you *please* take a look and merge? Every month you have breaking structural changes which make it difficult to rebase.

**Comment by danielgindi** [comment]
I've published this as `react-native-fast-image2`, as the package is now structure in a way that it cannot be directly installed from git.

**Comment by WrathChaos** [comment]
This repo should be merged on react-native-community. He simply does not care about the project or PR's. Thank you @danielgindi, awesome work man!

**Comment by t4ngth00** [comment]
@danielgindi Please rebase the package react-native-fast-image2 base on the latest version 8.3.4. version 8.3.3 introduced xcode 12 compatibility.

**Comment by danielgindi** [comment]
> @danielgindi Please rebase the package react-native-fast-image2 base on the latest version 8.3.4. Version 8.3.3 introduced the xcode 12 compatibility PR.

I have rebased and published

**Comment by deepakverma4** [comment]
@danielgindi I'm getting 
```
npm ERR! TAR_BAD_ARCHIVE: Unrecognized archive format
```
error when installing ```react-native-fast-image2```. Can you please look into it?

**Comment by danielgindi** [comment]
> @danielgindi I'm getting
> 
> ```
> npm ERR! TAR_BAD_ARCHIVE: Unrecognized archive format
> ```
> 
> error when installing `react-native-fast-image2`. Can you please look into it?

Re-released

**Comment by danielgindi** [comment]
@DylanVann what's up?

**Review by sanduluca — APPROVED** [review]

**Comment by gilbertl** [comment]
@DylanVann  it seems like this has been approved. Should I wait for this to be rebased and  merged? Or should I use react-native-fast-image-2 instead?

**Comment by gilbertl** [comment]
> > @danielgindi I'm getting
> > ```
> > npm ERR! TAR_BAD_ARCHIVE: Unrecognized archive format
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > error when installing `react-native-fast-image2`. Can you please look into it?
> 
> Re-released

We tried using  react-native-fast-image2 and ran into some issues on Android.

In the release builds, loading any image with defaultSource will crash the app on first load. Subsequent opens of the app will not result in this crash.

We were setting defaultSource like this

```
defaultSource={require('../../../assets/images/square-placeholder.png')}
```

It seems to work fine on iOS.

**Comment by danielgindi** [comment]
@gilbertl 
Can you test with a debug build? 
Maybe even put some breakpoints where the code gives Glide the url?
Have you tested with a different image?
So you by any chance have any special circumstances like perhaps splitting the apk and downloading the assets in the background?

**Comment by danielgindi** [comment]
> @danielgindi if it isn't too much trouble could you please rebase `react-native-fast-image2`?

Done.

@DylanVann please, people really need this.

**Comment by KingAmo** [comment]
any news here?

**Comment by mohandere** [comment]
> any news here?

Hey @DylanVann could please have a look on this?

**Comment by zholmes1** [comment]
Waiting patiently

**Comment by arnekolja** [comment]
Working like a charm on iOS, but on Android the item stays empty until it is loaded. Not crashing though.

**Comment by fukemy** [comment]
@danielgindi 
please look at this issues, and open issue mode of your project for question
> Working like a charm on iOS, but on Android the item stays empty until it is loaded. Not crashing though.

**Comment by fukemy** [comment]
Screen Shot 2022-06-16 at 13 24 47
i got this error after provide defaultSource props

**Comment by danielgindi** [comment]
> Screen Shot 2022-06-16 at 13 24 47
> 
> i got this error after provide defaultSource props

Remove the original lib.

**Comment by fukemy** [comment]
@danielgindi  your Pr work as expected on IOS, but i dont have time to test Android before removed some conflicts from dependency.
 I dont know why @DylanVann doesnt care about this topic.

**Comment by DylanVann** [comment]
Sorry for the delay. Thank you for the work on this. I’ve merged some work from #550, #741, and #498. If what’s on main now is missing anything important please open another PR. Thank you!

**Comment by fukemy** [comment]
thanks @DylanVann, now it's working

@danielgindi, does defaultSource work with animating image? I want to add a gif as placeholder, but it's does not animate when showing

**Pull request closed without merging**